source/common/common/matchers.cc

```diff
@@ -111,5 +113,124 @@ bool MetadataMatcher::match(const envoy::api::v2::core::Metadata& metadata) cons
+    value_ = config.prefix_match();
+    break;
+  case envoy::type::matcher::HeaderMatcher::kSuffixMatch:
+    header_match_type_ = HeaderMatchType::Suffix;
+    value_ = config.suffix_match();
+    break;
+  case envoy::type::matcher::HeaderMatcher::HEADER_MATCH_SPECIFIER_NOT_SET:
+    FALLTHRU;
+  default:
+    header_match_type_ = HeaderMatchType::Present;
+    break;
+  }
+}
+
+HeaderUtility::HeaderData::HeaderData(const Json::Object& config)
+    : HeaderData([&config] {
+        envoy::type::matcher::HeaderMatcher header_matcher;
+        Envoy::Config::RdsJson::translateHeaderMatcher(config, header_matcher);
+        return header_matcher;
+      }()) {}
+
+bool HeaderUtility::matchHeaders(const Http::HeaderMap& request_headers,
+                                 const std::vector<HeaderData>& config_headers) {
+  // TODO (rodaine): Should this really allow empty headers to always match?
+  if (!config_headers.empty()) {
+    for (const HeaderData& cfg_header_data : config_headers) {
+      if (!matchHeaders(request_headers, cfg_header_data)) {
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+bool HeaderUtility::matchHeaders(const Http::HeaderMap& request_headers,
+                                 const HeaderData& header_data) {
+  const Http::HeaderEntry* header = request_headers.get(header_data.name_);
+
+  if (header == nullptr) {
+    return header_data.invert_match_ && header_data.header_match_type_ == HeaderMatchType::Present;
+  }
+
+  bool match;
+  switch (header_data.header_match_type_) {
+  case HeaderMatchType::Value:
+    match = header_data.value_.empty() || header->value() == header_data.value_.c_str();
+    break;
+  case HeaderMatchType::Regex:
+    match = std::regex_match(header->value().c_str(), header_data.regex_pattern_);
+    break;
+  case HeaderMatchType::Range: {
+    int64_t header_value = 0;
+    match = StringUtil::atol(header->value().c_str(), header_value, 10) &&
+            header_value >= header_data.range_.start() && header_value < header_data.range_.end();
+    break;
+  }
+  case HeaderMatchType::Present:
+    match = true;
+    break;
+  case HeaderMatchType::Prefix:
+    match = absl::StartsWith(header->value().getStringView(), header_data.value_);
+    break;
+  case HeaderMatchType::Suffix:
+    match = absl::EndsWith(header->value().getStringView(), header_data.value_);
+    break;
+  default:
+    NOT_REACHED_GCOVR_EXCL_LINE;
+  }
+
+  return match != header_data.invert_match_;
+}
+
+void HeaderUtility::addHeaders(Http::HeaderMap& headers, const Http::HeaderMap& headers_to_add) {
+  headers_to_add.iterate(
+      [](const Http::HeaderEntry& header, void* context) -> Http::HeaderMap::Iterate {
+        Http::HeaderString k;
+        k.setCopy(header.key().c_str(), header.key().size());
+        Http::HeaderString v;
+        v.setCopy(header.value().c_str(), header.value().size());
+        static_cast<Http::HeaderMapImpl*>(context)->addViaMove(std::move(k), std::move(v));
```

Comment:
I'd like to not depend on the `common/http` package and I believe that means using the `HeaderMap` interface here -- any objections to changing this to do that?

Reply:
I believe you can use addCopy in HeaderMap, and pass the `HeaderString`s directly from `header` here, this would avoid one copy above so it would be same. You might want to change the second parameter of addCopy from `const string&` to `absl::string_view` so it would avoid copying the value.

Reply:
`addCopy` takes an `Http::LowerCaseString` for key have to construct one in a similar way but did change value to be an `absl::string_view`.

Reply:
Ah I missed that HeaderString -> LowerCaseString part, then probably it is better to add `addViaMove` to the interface itself? I would prefer make those change in another PR and get it merged first.

Reply:
sure -- i can definitely do that